src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfo.Cache.cs

```diff
@@ -582,18 +587,34 @@ internal void InitializePropCache()
             }
 
             JsonPropertyInfo[] array = PropInitFunc(context);
-            var properties = new JsonPropertyDictionary<JsonPropertyInfo>(Options.PropertyNameCaseInsensitive, array.Length);
+            Dictionary<string, JsonPropertyInfo>? ignoredMembers = null;
+            JsonPropertyDictionary<JsonPropertyInfo> propertyCache = new(Options.PropertyNameCaseInsensitive, array.Length);
+
             for (int i = 0; i < array.Length; i++)
             {
-                JsonPropertyInfo property = array[i];
-                if (!properties.TryAdd(property.NameAsString, property))
+                JsonPropertyInfo jsonPropertyInfo = array[i];
+                bool hasJsonInclude = jsonPropertyInfo.SrcGen_HasJsonInclude;
+
+                if (!jsonPropertyInfo.SrcGen_IsPublic)
+                {
+                    if (hasJsonInclude)
```

Comment:
Could this be detected during source generation instead of at runtime? (and avoid passing this info over)

Reply:
Yes this a good idea, but makes testing a bit too difficult to tackle in this PR at this point. Switching from runtime exceptions to compile-time warnings means that test projects won't compile & we'd have to test using a different approach. Will tackle in a follow-up.

Reply:
Hm I think we might have to keep this functionality as-is, given that some of the things that cause property-name conflicts (e.g. naming policy) are specified at runtime only. I'll make sure.